worker/caasrbacmapper/mapper.go

```diff
@@ -0,0 +1,182 @@
+	// with a given kubernetes service account UID. If no result is found
+	// errors.NotFound is returned. All other errors should be considered
+	// internal to the interface operation.
+	AppNameForServiceAccount(types.UID) (string, error)
+}
+
+// MapperWorker is a Mapper that also implements the worker interface
+type MapperWorker interface {
+	Mapper
+	worker.Worker
+}
+
+// DefaultMapper is a default implementation of the MapperWorker interface. It's
+// responsible for watching ServiceAccounts on a given model and
+type DefaultMapper struct {
+	catacomb     catacomb.Catacomb
+	lock         *sync.RWMutex
+	logger       Logger
+	saInformer   core.ServiceAccountInformer
+	saNameUIDMap map[string]types.UID
+	saUIDAppMap  map[types.UID]string
+	workQueue    workqueue.RateLimitingInterface
+}
+
+// AppNameForServiceAccount implements Mapper interface
+func (d *DefaultMapper) AppNameForServiceAccount(id types.UID) (string, error) {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	appName, found := d.saUIDAppMap[id]
+	if !found {
+		return "", errors.NotFoundf("no service account for app found with id %v", id)
+	}
+	return appName, nil
+}
+
+func (d *DefaultMapper) enqueueServiceAccount(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		d.logger.Errorf("failed enqueuing service account: %v", err)
+		return
+	}
+	d.workQueue.Add(key)
+}
+
+// Kill implements Worker interface
+func (d *DefaultMapper) Kill() {
+	d.catacomb.Kill(nil)
+}
+
+func (d *DefaultMapper) loop() error {
+	defer d.workQueue.ShutDown()
+	d.saInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    d.enqueueServiceAccount,
+		DeleteFunc: d.enqueueServiceAccount,
+		UpdateFunc: func(_, newObj interface{}) {
+			d.enqueueServiceAccount(newObj)
+		},
+	})
+
+	go d.saInformer.Informer().Run(d.catacomb.Dying())
+
+	// Wait until runs the below for loop every one second until the catacomb
+	// dies. The for loop processes all items in the queue till it's empty and
+	// the cycle repeats. This is to stop checking thrashing about and is the
+	// prescribed k8s way to process.
+	go wait.Until(func() {
+		for d.processNextQueueItem() {
+		}
+	}, time.Second, d.catacomb.Dying())
+
+	select {
+	case <-d.catacomb.Dying():
+	}
+
+	return nil
+}
+
+// NewMapper constructs a new DefaultMapper for the supplied logger and
+// ServiceAccountInformer
+func NewMapper(logger Logger, informer core.ServiceAccountInformer) (*DefaultMapper, error) {
+	dm := &DefaultMapper{
+		lock:         new(sync.RWMutex),
+		logger:       logger,
+		saInformer:   informer,
+		saNameUIDMap: map[string]types.UID{},
+		saUIDAppMap:  map[types.UID]string{},
+		workQueue:    workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &dm.catacomb,
+		Work: dm.loop,
+	}); err != nil {
+		return dm, errors.Trace(err)
+	}
+	return dm, nil
+}
+
+func (d *DefaultMapper) processNextQueueItem() bool {
```

Comment:
`processNextQueueItem` can only return `false` if `shutdown ==true` if I wasn't wrong.
In any other case, it should return `true`.
Because we don't want any unexpected event to kill this worker.

Reply:
I am not sure what you mean here. Can you expand. Standard k8s practice agrees with what we have done.

Reply:
https://github.com/kubernetes/sample-controller/blob/master/controller.go#L187
I was reading here.

Reply:
Ah yep I see now. Thanks will fix it up.